### PR TITLE
gradients for depthwise 2d convolutions

### PR DESCRIPTION
**Pull request by asross**
#### Description

This PR addresses https://github.com/tensorflow/tfjs/issues/71 by introducing gradient support for `depthwiseConv2D` (letting users train or compute saliency maps for models using depthwise convolutions).

Gradient functions are very similar to the corresponding functions for regular 2D convolutions (and actually return the same results in the special case that no further transformations of the output occur), except instead of iterating over all the output channels, we only iterate over a subset. The indexing into filter weights also becomes different than the indexing into outputs.

In implementing this, I found [reference](https://github.com/tensorflow/tensorflow/blob/51ef16057b4625e0a3e2943a9f1bbf856cf098ca/tensorflow/core/kernels/depthwise_conv_grad_op.cc#L940-L986) [implementations](https://github.com/tensorflow/tensorflow/blob/51ef16057b4625e0a3e2943a9f1bbf856cf098ca/tensorflow/core/kernels/depthwise_conv_grad_op.cc#L459-L513) in C++ very helpful.

---

##### For repository owners only:

Please remember to apply all applicable tags to your pull request.
Tags: FEATURE, BREAKING, BUG, PERF, DEV, DOC, SECURITY

For more info see: https://github.com/tensorflow/tfjs/blob/master/DEVELOPMENT.md

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1069)

### CODE REVIEW

File: src/kernels/backend_cpu.ts [hunk 1]
```diff
@@ -1193,6 +1193,116 @@ export class MathBackendCPU implements KernelBackend {
...
+              const wR = yR * strideHeight - xRCorner;
+
+              for (let yC = xCMin; yC < yCMax; ++yC) {
+                const wC = yC * strideWidth - xCCorner;
+                const dyOffset = dyS0 * b + dyS1 * yR + dyS2 * yC;
+                const fltOffset = fltS0 * (filterHeight - 1 - wR) +
+                    fltS1 * (filterWidth - 1 - wC) + fltS2 * d1;
+
+                const d2Min = d1 * chMul;
+                const d2Max = d2Min + chMul;
```
Comment (asross) [inline src/kernels/backend_cpu.ts]: This selects only the output channels corresponding to the current gradient component's input channel.

File: src/kernels/backend_cpu.ts [hunk 2]
```diff
@@ -1193,6 +1193,116 @@ export class MathBackendCPU implements KernelBackend {
...
+                for (let d2 = d2Min; d2 < d2Max; ++d2) {
+                  const dm = d2 - d2Min;
+                  const pixel = dyValues[dyOffset + d2];
+                  const weight = fltValues[fltOffset + dm];
```
Comment (asross) [inline src/kernels/backend_cpu.ts]: `dm` just goes from `0` to `chMul`. It might actually be more intuitive to index that way (so `d2 = d1 * chMul + dm`) -- let me know if you'd prefer that.

File: src/ops/conv2d_depthwise_test.ts [hunk 3]
```diff
@@ -372,3 +372,74 @@ describeWithFlags('depthwiseConv2D', ALL_ENVS, () => {
...
+      [[15,15],[16,16],[12,12]],
+      [[7,7], [8,8], [9,9]]
+    ], [
+      [[8,8], [9,9], [6,6]],
+      [[4,4], [5,5], [4,4]]
+    ]])
+
+    expectArraysClose(grad, expectedGrad);
+  });
+});
```
Comment (asross) [inline src/ops/conv2d_depthwise_test.ts]: Let me know if you think this test is too complicated / not complicated enough. I basically just copied outputs from [this notebook](https://colab.research.google.com/drive/1tqSds-WqDLC_IZ4sVAj1sR-cfQlvczxW) into the test.

File: src/ops/conv.ts [hunk 4]
```diff
@@ -479,12 +479,106 @@ export class ConvOps {
...
+          util.isInt(pad as number),
+          `Error in depthwiseConv2DDerInput: pad must be an integer when using, ` +
+              `dimRoundingMode ${dimRoundingMode} but got pad ${pad}.`);
+    }
+
+    const dilations = 1;
+
+    const convInfo = conv_util.computeConv2DInfo(
+        xShape4D, filter.shape, strides, dilations, pad, dimRoundingMode,
+        true /* depthwise */);
```
Comment (asross) [inline src/ops/conv.ts]: For a while, I was missing this `true`, which ended up being a very time-consuming/confusing bug. Might 
 be a sign that the interface for `computeConv2DInfo` is getting a bit clunky, though I know it's not part of the public API...

**Comment by dsmilkov** [comment]
Really really nice work. Left tiny comments - feel free to go ahead with WebGL

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[src/kernels/backend_cpu.ts, line 1247 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1069#-LDmj5esZWZLrL_whyE5-r1-1247:-LDq_YPS2peGEOB1w3NE:b-83sjoh) ([raw file](https://github.com/tensorflow/tfjs-core/blob/317a67c0eb3d02f26b61ec8c4f7db575d2a7f069/src/kernels/backend_cpu.ts#L1247)):*
<details><summary><i>Previously, asross (Andrew Ross) wrote…</i></summary><blockquote>

This selects only the output channels corresponding to the current gradient component's input channel.
</blockquote></details>

Nice!

---

*[src/kernels/backend_cpu.ts, line 1251 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1069#-LDmj5esZWZLrL_whyE5-r1-1251:-LDq_T6oAeICYthEdkLX:b251hx4) ([raw file](https://github.com/tensorflow/tfjs-core/blob/317a67c0eb3d02f26b61ec8c4f7db575d2a7f069/src/kernels/backend_cpu.ts#L1251)):*
<details><summary><i>Previously, asross (Andrew Ross) wrote…</i></summary><blockquote>

`dm` just goes from `0` to `chMul`. It might actually be more intuitive to index that way (so `d2 = d1 * chMul + dm`) -- let me know if you'd prefer that.
</blockquote></details>

Yeah, it would be a bit cleaner to do dm from 0  to chMul.

---

*[src/ops/conv.ts, line 578 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1069#-LDmj5esZWZLrL_whyE7-r1-578:-LDqaCC74_Cwnkmo7eD_:bg6oqnq) ([raw file](https://github.com/tensorflow/tfjs-core/blob/317a67c0eb3d02f26b61ec8c4f7db575d2a7f069/src/ops/conv.ts#L578)):*
<details><summary><i>Previously, asross (Andrew Ross) wrote…</i></summary><blockquote>

For a while, I was missing this `true`, which ended up being a very tricky bug. Might 
 be a sign that the interface for `computeConv2DInfo` is getting a bit clunky, though I know it's not part of the public API...
</blockquote></details>

yeah, it gradually evolved to be all encompassing and splitting it into different entry points is a great suggestion, but as you pointed out correctly, may not be worth the effort for now.

---

*[src/ops/conv.ts, line 549 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1069#-LDq_m_J9UHArH6J3Kf9:-LDq_m_J9UHArH6J3KfA:bso9egq) ([raw file](https://github.com/tensorflow/tfjs-core/blob/2f95f00b514398125f5898016d03e70f5ee60fa0/src/ops/conv.ts#L549)):*
> ```XML
>     util.assert(
>         xShape4D.length === 4,
>         `Error in depthwiseConv2DDerInput: inShape must be length 4, but got length ` +
> ```

fix wrap

---

*[src/ops/conv.ts, line 571 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1069#-LDq_pxF9gJZY7A7pa25:-LDq_pxF9gJZY7A7pa26:bso9egq) ([raw file](https://github.com/tensorflow/tfjs-core/blob/2f95f00b514398125f5898016d03e70f5ee60fa0/src/ops/conv.ts#L571)):*
> ```XML
>       util.assert(
>           util.isInt(pad as number),
>           `Error in depthwiseConv2DDerInput: pad must be an integer when using, ` +
> ```

fix wrap

---

*[src/ops/conv.ts, line 624 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1069#-LDqahwz8LIjux1n8mVu:-LDqahwz8LIjux1n8mVv:bb7r9an) ([raw file](https://github.com/tensorflow/tfjs-core/blob/2f95f00b514398125f5898016d03e70f5ee60fa0/src/ops/conv.ts#L624)):*
> ```XML
>     util.assert(
>         x4D.rank === 4,
>         `Error in depthwiseConv2DDerFilter: input must be rank 4, but got shape ` +
> ```

wrap

---

*[src/ops/conv.ts, line 632 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1069#-LDqaiXBAGbxJhsk-iii:-LDqaiXBAGbxJhsk-iij:bb7r9an) ([raw file](https://github.com/tensorflow/tfjs-core/blob/2f95f00b514398125f5898016d03e70f5ee60fa0/src/ops/conv.ts#L632)):*
> ```XML
>     util.assert(
>         filterShape.length === 4,
>         `Error in depthwiseConv2DDerFilter: filterShape must be length 4, but got ` +
> ```

wrap

---

*[src/ops/conv.ts, line 646 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/1069#-LDqag922owj5BORnBZr:-LDqag922owj5BORnBZs:bso9egq) ([raw file](https://github.com/tensorflow/tfjs-core/blob/2f95f00b514398125f5898016d03e70f5ee60fa0/src/ops/conv.ts#L646)):*
> ```XML
>       util.assert(
>           util.isInt(pad as number),
>           `Error in depthwiseConv2DDerFilter: pad must be an integer when using, ` +
> ```

fix wrap

---

*[src/ops/conv2d_depthwise_test.ts, line 445 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1069#-LDmj5esZWZLrL_whyE8-r1-445:-LDqamZ93_Hmj507v1zM:b-1okvdx) ([raw file](https://github.com/tensorflow/tfjs-core/blob/317a67c0eb3d02f26b61ec8c4f7db575d2a7f069/src/ops/conv2d_depthwise_test.ts#L445)):*
<details><summary><i>Previously, asross (Andrew Ross) wrote…</i></summary><blockquote>

Let me know if you think this test is too complicated / not complicated enough. I basically just copied outputs from [this notebook](https://colab.research.google.com/drive/1tqSds-WqDLC_IZ4sVAj1sR-cfQlvczxW) into the test.
</blockquote></details>

This test is great. Though I would write a couple more that test different scenarious, e.g. stride = 2, chMul = 2, another padding type.  Don't worry about tests being mostly repetitive (we don't ship these) - the more self containing the test is - the better.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1069#-:-LDqbDs9-AvSm5RTYuS2:bjo7x5u)*

**Comment by dsmilkov** [comment]
Really really nice work! Left a few comments, but they are all very small!

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[src/kernels/webgl/conv_backprop_gpu_depthwise.ts, line 3 at r3](https://reviewable.io/reviews/tensorflow/tfjs-core/1069#-LDvo4jOCPZM4B2zFLFj:-LDvo4jPDJPUGNji7tGU:bn3g4qa) ([raw file](https://github.com/tensorflow/tfjs-core/blob/013ef5dcd5c54373305a5697e54cca7536f4021d/src/kernels/webgl/conv_backprop_gpu_depthwise.ts#L3)):*
> ```XML
> /**
>  * @license
>  * Copyright 2017 Google Inc. All Rights Reserved.
> ```

2018 Google LLC

---

*[src/kernels/webgl/conv_backprop_gpu_depthwise.ts, line 46 at r3](https://reviewable.io/reviews/tensorflow/tfjs-core/1069#-LDvpA7k3yvU1FEALjde:-LDvpA7k3yvU1FEALjdf:bn8vixp) ([raw file](https://github.com/tensorflow/tfjs-core/blob/013ef5dcd5c54373305a5697e54cca7536f4021d/src/kernels/webgl/conv_backprop_gpu_depthwise.ts#L46)):*
> ```XML
>         float dotProd = 0.0;
> 
>         for (int b = 0; b < ${convInfo.batchSize}; b++) {
> ```

add a `// TODO: Vec4 over the batch size` . (will chat in person after this PR is in)

---

*[src/kernels/webgl/conv_backprop_gpu_depthwise.ts, line 124 at r3](https://reviewable.io/reviews/tensorflow/tfjs-core/1069#-LDvowz8-52thFMOlVf2:-LDvowz8-52thFMOlVf3:b-sofwqr) ([raw file](https://github.com/tensorflow/tfjs-core/blob/013ef5dcd5c54373305a5697e54cca7536f4021d/src/kernels/webgl/conv_backprop_gpu_depthwise.ts#L124)):*
> ```XML
>             int wCPerm = ${filterWidth} - 1 - wC;
> 
>             for (int dm = 0; dm < ${channelMul}; dm++) {
> ```

add a `// TODO: Vec4 over the channelMul` (we'll chat in person after this PR)

---

*[src/ops/conv.ts, line 508 at r3](https://reviewable.io/reviews/tensorflow/tfjs-core/1069#-LDvpWPs2qUL9GuRsDUk:-LDvpWPs2qUL9GuRsDUl:b-s55ajn) ([raw file](https://github.com/tensorflow/tfjs-core/blob/013ef5dcd5c54373305a5697e54cca7536f4021d/src/ops/conv.ts#L508)):*
> ```XML
>   }
> 
>   static depthwiseConv2dDerInput<T extends Tensor3D|Tensor4D>(
> ```

since these 2 private methods don't have any decorators, move them to the bottom of the file and remove `static` keyword. (context: decorators are only allowed on class methods, and not free-floating functions, thus we have these strange looking classes with bunch of static methods )

---

*[src/ops/conv2d_depthwise_test.ts, line 433 at r3](https://reviewable.io/reviews/tensorflow/tfjs-core/1069#-LDvreIAC5i6TChlmiWL:-LDvreIAC5i6TChlmiWM:bc504fd) ([raw file](https://github.com/tensorflow/tfjs-core/blob/013ef5dcd5c54373305a5697e54cca7536f4021d/src/ops/conv2d_depthwise_test.ts#L433)):*
> ```XML
>   it('wrt input, squared output', () => {
>     const grad = tf.grad(
>       (x: tf.Tensor4D) => tf.square(tf.depthwiseConv2d(x, filter, stride, pad))
> ```

tip: https://js.tensorflow.org/api/0.11.2/#grad
tf.grad returns a function with an optional second argument being dy
so you can specify your own dy
e.g. https://github.com/tensorflow/tfjs-core/blob/master/src/ops/binary_ops_test.ts#L773

so it might be simpler to just use that API instead of calling tf.square()

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1069#-:-LDvroDg1ty7zK8Z3VPA:b-sjlrza)*

**Comment by dsmilkov** [comment]
:lgtm_strong: Wohoo!

---

Review status: 4 of 6 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1069#-:-LDwIIhU8hTXqfbh9geY:b-2k0akc)*
